### PR TITLE
Introduce blobbasefee

### LINKED ISSUE #14742: Support for BLOBBASEFEE from EIP 7516
Part of https://github.com/ethereum/solidity/issues/14739
First step:

- [x] Introduce the blobbasefee opcode as specified in https://eips.ethereum.org/EIPS/eip-7516

Second step:

- Introduce ``block.blobbasefee`` backed by the above opcode.

Use https://github.com/ethereum/solidity/pull/11647 as reference.

Also needs to take care of this not being breaking: https://github.com/ethereum/solidity/pull/11763

### PR DESCRIPTION
**Pull request by nikola-matic**
Closes: https://github.com/ethereum/solidity/issues/14742
~~Depends on: https://github.com/ethereum/solidity/pull/14743~~

### CODE REVIEW

File: libyul/backends/evm/EVMDialect.cpp [hunk 1]
```diff
@@ -120,6 +120,13 @@ std::set<YulString> createReservedIdentifiers(langutil::EVMVersion _evmVersion)
 		return _instr == evmasm::Instruction::BASEFEE && _evmVersion < langutil::EVMVersion::london();
 	};
 
+	// TODO remove this in 0.9.0. We allow creating functions or identifiers in Yul with the name
```
Comment (nikola-matic) [inline libyul/backends/evm/EVMDialect.cpp:123]: We have exceptions for `basefee`, `blobbasefee` and `prevrandao`, which need to be removed for `0.9.0`. Do we have an issue tracking this?

Comment (r0qs) [reply libyul/backends/evm/EVMDialect.cpp:123]: For the `prevrandao` I believe is this one here: https://github.com/ethereum/solidity/issues/13886 but we don't have for the others.

File: libyul/backends/evm/EVMDialect.cpp [hunk 2]
```diff
@@ -120,6 +120,13 @@ std::set<YulString> createReservedIdentifiers(langutil::EVMVersion _evmVersion)
...
+	// blobbasefee for VMs before cancun.
+	auto blobBaseFeeException = [&](evmasm::Instruction _instr) -> bool
+	{
+		return _instr == evmasm::Instruction::BLOBBASEFEE && _evmVersion < langutil::EVMVersion::cancun();
+	};
```
Comment (r0qs) [inline libyul/backends/evm/EVMDialect.cpp:128]: Maybe it would be nice a little refactor, not in this PR though. Because we will now add more three exception annonymous functions here for the cancun support. So we could at least have a proper function to create such opcode exceptions, since they are basically the same. Only `prevrandao` has a different logic.

File: libevmasm/Instruction.cpp [hunk 3]
```diff
@@ -230,6 +231,7 @@ static std::map<Instruction, InstructionInfo> const c_instructionInfo =
 	{ Instruction::CHAINID,		{ "CHAINID",		0, 0, 1, false, Tier::Base } },
 	{ Instruction::SELFBALANCE,	{ "SELFBALANCE",	0, 0, 1, false, Tier::Low } },
 	{ Instruction::BASEFEE,     { "BASEFEE",        0, 0, 1, false, Tier::Base } },
+	{ Instruction::BLOBBASEFEE,     { "BLOBBASEFEE",        0, 0, 1, false, Tier::Base } },
```
Comment (nikola-matic) [inline libevmasm/Instruction.cpp]:
```suggestion
	{ Instruction::BLOBBASEFEE, { "BLOBBASEFEE",        0, 0, 1, false, Tier::Base } },
```

File: docs/cheatsheet.rst [hunk 4]
```diff
@@ -56,14 +56,15 @@ Members of ``address``
...
-.. index:: blockhash, block, block;basefree, block;chainid, block;coinbase, block;difficulty, block;gaslimit, block;number, block;prevrandao, block;timestamp
+.. index:: blockhash, block, block;basefee, block;blobbasefee, block;chainid, block;coinbase, block;difficulty, block;gaslimit, block;number, block;prevrandao, block;timestamp
 .. index:: gasleft, msg;data, msg;sender, msg;sig, msg;value, tx;gasprice, tx;origin
 
 Block and Transaction Properties
 ================================
 
 - ``blockhash(uint blockNumber) returns (bytes32)``: hash of the given block - only works for 256 most recent blocks
 - ``block.basefee`` (``uint``): current block's base fee (`EIP-3198 <https://eips.ethereum.org/EIPS/eip-3198>`_ and `EIP-1559 <https://eips.ethereum.org/EIPS/eip-1559>`_)
+- ``block.blobbasefee`` (``uint``): current block's blob base fee (`EIP-7516 <https://eips.ethereum.org/EIPS/eip-7516>`_ and `EIP-4844 <https://eips.ethereum.org/EIPS/eip-4844>`_)
```
Comment (r0qs) [inline docs/cheatsheet.rst:67]: Maybe we should just mention the blobbasefee eip here (i.e. eip-7516) because the other will be mentioned in the other PR, eip-4844.
```suggestion
- ``block.blobbasefee`` (``uint``): current block's blob base fee (`EIP-7516 <https://eips.ethereum.org/EIPS/eip-7516>`_ )
```

File: libyul/backends/evm/EVMDialect.cpp (hunk 1 above)
Comment (cameel) [reply libyul/backends/evm/EVMDialect.cpp:123]: #13886 does not mention this hack though. I'm not sure we'd even remember to remove this when fixing it.

But anyway, since this is not the last opcode we're adding, we should just create a catch-all issue for Yul identifiers we're planning to reserve in the next breaking release. Include `difficulty` and `prevrandao` there too.

Comment (nikola-matic) [reply libyul/backends/evm/EVMDialect.cpp:123]: Yup, this is why I asked - I wanted to create one before  new years, but was unsure as to whether we already had one laying around.

File: test/EVMHost.cpp [hunk 5]
```diff
@@ -136,6 +136,7 @@ EVMHost::EVMHost(langutil::EVMVersion _evmVersion, evmc::VM& _vm):
 	tx_context.chain_id = evmc::uint256be{1};
 	// The minimum value of basefee
 	tx_context.block_base_fee = evmc::bytes32{7};
+	tx_context.blob_base_fee = evmc::bytes32{7};
```
Comment (cameel) [inline test/EVMHost.cpp]: I think that for blob base fee the minimum is 1, not 7. For base fee it's 7 [due to integer math used](https://github.com/ethereum/solidity/pull/11647/files#r685037191), but blob base fee uses a different formula. `fake_exponential(MIN_BLOB_GASPRICE, header.excess_blob_gas, BLOB_GASPRICE_UPDATE_FRACTION)` will give you 1 when `excess_blob_gas` is 0.

That aside from why we're setting it to the minimum value in the first place - perhaps 7 is fine anyway, these values seem pretty arbitrary, but the comment makes it seem like it covers both and 7 was chosen due to being the minimum. I'd either change it to 1 or add a comment saying why it's 7.

Same for `Interpreter.h`.

File: test/libsolidity/syntaxTests/inlineAssembly/blobbasefee_reserved_cancun.sol [hunk 6]
```diff
@@ -0,0 +1,12 @@
+contract C {
+    function f() public view returns (uint ret) {
+        assembly {
+            let blobbasefee := sload(0)
+            ret := blobbasefee
+        }
+    }
+}
+// ====
+// EVMVersion: =cancun
```
Comment (cameel) [inline test/libsolidity/syntaxTests/inlineAssembly/blobbasefee_reserved_cancun.sol]: It will still be an error on later versions, won't it?
```suggestion
// EVMVersion: >=cancun
```

File: test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+        assembly {
+            ret := blobbasefee()
+        }
+    }
+}
+// ====
+// EVMVersion: <=shanghai
+// ----
+// TypeError 1006: (74-91): "blobbasefee" is not supported by the VM version.
+// DeclarationError 4619: (187-198): Function "blobbasefee" not found.
```
Comment (cameel) [inline test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol]: Something must be wrong here. I'd expect this test to trigger error 6679 (`only available for Cancun-compatible VMs`), but it instead reports 4619 (`Function ... not found`). `blobbasefee` should be recognized as a built-in even if it's rejected, just as it is for `basefee`.

Note that this does not happen for `basefee`. If you replace `blobbasefee()` with `basefee()` and run this test with `--evm-version berlin`, it generates error 5430 (`only available for London-compatible VMs`) as expected.

This discrepancy on its own is not a big issue, but if it's a bug, something else could be broken so I think it's better to check why this opcode is handled differently all of a sudden.

And by the way, since 5430 seems reproducible, we could add back the corresponding case for it too and have one less code in `error-codes.py`. The coverage was removed in #11763, but I'm not sure why. This makes me wonder if this isn't something weird on my side and both actually report 4619, but I rechecked multiple times, even with a release binary, and I can trigger this error every time.

File: Changelog.md [hunk 8]
```diff
@@ -1,6 +1,8 @@
 ### 0.8.24 (unreleased)
 
 Language Features:
+ * Introduce global ``block.blobbasefee`` for retrieving the blob base fee of the current block.
+ * Yul: Introduce builtin ``blobbasefee()`` for retrieving the blob base fee of the current block.
```
Comment (cameel) [inline Changelog.md:5]: Cancun related changes should also be listed under [Target Options](https://docs.soliditylang.org/en/latest/using-the-compiler.html#target-options) in the docs.

File: libyul/backends/evm/EVMDialect.cpp (hunk 1 above)
Comment (ekpyron) [reply libyul/backends/evm/EVMDialect.cpp:123]: We may have parts of this stuff lingering on the ``breaking`` branch already - but feel free to create issues for this and mark them ``must have eventually`` and ``breaking`` and add them to the 0.9 milestone and such, if you haven't already.

File: test/tools/yulInterpreter/EVMInstructionInterpreter.cpp
Comment (cameel) [inline test/tools/yulInterpreter/EVMInstructionInterpreter.cpp:1]: Just noticed there are no pure Yul tests (`test/libyul/`) in the PR. Please take a look at #14757 and add some.

File: libyul/backends/evm/EVMDialect.cpp (hunk 1 above)
Comment (nikola-matic) [reply libyul/backends/evm/EVMDialect.cpp:123]: [Done](https://github.com/ethereum/solidity/issues/14770).

File: docs/cheatsheet.rst (hunk 4 above)
Comment (nikola-matic) [reply docs/cheatsheet.rst:67]: From what I've seen, the two are related, and we generally mention both in such cases.

File: test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol (hunk 7 above)
Comment (nikola-matic) [reply test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol]: For `basefee` actually, the assembly block was explicitly removed - https://github.com/ethereum/solidity/pull/11763/files#diff-77a7a77f2a2831156ab9f1652dabc5af0858e8ff6739dd3890540ff1d3c5f875

Comment (nikola-matic) [reply test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol]: OK, so of course - using the `<=shanghai` instead of `=shanghai` will produce a different error for each VM (only available for XXX-compatible), so that's clearly not gonna work in our tests that are run across different EVM versions (which is why Hari removed it in the first place). It is however weird that this doesn't throw the error you mentioned, but triggers a declaration error instead. I'll look into it.

Comment (nikola-matic) [reply test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol]: I'll leave this as is, and once we make `cancun` the default version in the tests (and compiler), the test should work just fine; this will also have to be set to `=shanghai` due to the error taking into account the fork name, and therefore changing the output depending on the selected EVM version.

Comment (cameel) [reply test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol]: >  that's clearly not gonna work in our tests that are run across different EVM versions (which is why Hari removed it in the first place)

I don't think this was the reason. The test was just fine with `=shanghai`. That was much better than having an uncovered error that we have to explicitly ignore in `error_codes.py`. I still think we should restore that test to the previous form.

Comment (nikola-matic) [reply test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol]: Yup, I'll pin this to `=shanghai`, but regardless, once we make `cancun` the default, the test will have to be updated since it will yield another error (only available with `cancun` and you're compiling with `xxxxx`).

File: libyul/backends/evm/EVMDialect.cpp (hunk 2 above)
Comment (nikola-matic) [reply libyul/backends/evm/EVMDialect.cpp:128]: I thought about doing exactly this, but didn't wanna introduce yet another thing to review, especially seeing as we'll be releasing `0.9.0` sometime this year, and this will get removed anyway. As in - we could certainly refactor this to look nicer, but I don't think it's really worthwhile since it's gonna get tossed anyway.

File: docs/using-the-compiler.rst [hunk 9]
```diff
@@ -176,6 +176,8 @@ at each version. Backward compatibility is not guaranteed between each version.
    - Introduces ``prevrandao()`` and ``block.prevrandao``, and changes the semantics of the now deprecated ``block.difficulty``, disallowing ``difficulty()`` in inline assembly (see `EIP-4399 <https://eips.ethereum.org/EIPS/eip-4399>`_).
 - ``shanghai`` (**default**)
    - Smaller code size and gas savings due to the introduction of ``push0`` (see `EIP-3855 <https://eips.ethereum.org/EIPS/eip-3855>`_).
+- ``cancun``
+   - he block's blob base fee (`EIP-7516 <https://eips.ethereum.org/EIPS/eip-7516>`_ and `EIP-4844 <https://eips.ethereum.org/EIPS/eip-4844>`_) can be accessed via the global ``block.blobbasefee`` or ``blobbasefee()`` in inline assembly.
```
Comment (cameel) [inline docs/using-the-compiler.rst]:
```suggestion
   - The block's blob base fee (`EIP-7516 <https://eips.ethereum.org/EIPS/eip-7516>`_ and `EIP-4844 <https://eips.ethereum.org/EIPS/eip-4844>`_) can be accessed via the global ``block.blobbasefee`` or ``blobbasefee()`` in inline assembly.
```

File: test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol [hunk 10]
```diff
@@ -0,0 +1,21 @@
+contract C {
+    function f() public view returns (uint ret) {
+        assembly {
+            let blobbasefee := sload(0)
```
Comment (r0qs) [inline test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol]: Should this be something like:
```suggestion
            sstore(0, 1000)
            let blobbasefee := sload(0)
```
Or just:
```suggestion
            let blobbasefee := 1000 // or any other value different from 0 and from the default blobbasefee value
```
Because the return of `sload(0)` will always be zero in that case, since nothing was stored at slot 0. I mean, not that this is really very relevant for the test, but just for consistency with function `g()`.

Comment (nikola-matic) [reply test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol]: Sure, makes sense.

File: test/libyul/yulSyntaxTests/blobbasefee_reserved_identifier_post_cancun.sol
Comment (cameel) [inline test/libyul/yulSyntaxTests/blobbasefee_reserved_identifier_post_cancun.sol:1]: It's a Yul file, use `.yul` extension. Same for other files you add for `libyul/`.

File: docs/units-and-global-variables.rst [hunk 11]
```diff
@@ -75,6 +75,7 @@ Block and Transaction Properties
 
 - ``blockhash(uint blockNumber) returns (bytes32)``: hash of the given block when ``blocknumber`` is one of the 256 most recent blocks; otherwise returns zero
 - ``block.basefee`` (``uint``): current block's base fee (`EIP-3198 <https://eips.ethereum.org/EIPS/eip-3198>`_ and `EIP-1559 <https://eips.ethereum.org/EIPS/eip-1559>`_)
+- ``block.blobbasefee`` (``uint``): current block's blob base fee (`EIP-7516 <https://eips.ethereum.org/EIPS/eip-7516>`_ and `EIP-4844 <https://eips.ethereum.org/EIPS/eip-4844>`_)
```
Comment (cameel) [inline docs/units-and-global-variables.rst:78]: Just noticed that both `basefee` and `blobbasefee` are missing in the index directive above.

File: test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol
Comment (cameel) [inline test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol:1]: For completeness we should also add `blobbasefee` to tests meant to cover all global variables/functions:
- `blobbasefee` to the `builtins` test in `FunctionCallGraph.cpp`.
- `environment_access` test in `ViewPureChecker.cpp`.
- `syntaxTests/viewPureChecker/inline_assembly_instructions_allowed...` group of tests

**Review by cameel — DISMISSED** [review]
I'm approving since it's good enough, but there are a few small things we could still tweak. See review comments.

We're missing SMTChecker support, but that's not essential for the release. Please just create an issue for it. The implementation would be as for `basefee` in #11785.

Fuzzer support is also missing and adding it seems fairly trivial (just a change in `ossfuzz/protoToYul.cpp`), but it's better left to @bshastry since he can verify it's done properly.

File: test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol (hunk 7 above)
Comment (cameel) [reply test/libsolidity/syntaxTests/types/magic_block_blobbasefee_error.sol]: But can you do this also for `basefee` (error 5430)?

**Comment by nikola-matic** [comment]
> I'm approving since it's good enough, but there are a few small things we could still tweak. See review comments.
> 
> We're missing SMTChecker support, but that's not essential for the release. Please just create an issue for it. The implementation would be as for `basefee` in #11785.
> 
> Fuzzer support is also missing and adding it seems fairly trivial (just a change in `ossfuzz/protoToYul.cpp`), but it's better left to @bshastry since he can verify it's done properly.

Yeah, saw the missing SMT support when initially implementing this last year (was especially evident when `grep`ing `basefee`); I may have even asked (or was planning to and then didn't wanna ping people during Christmas) whose responsibility that is. But yeah, given that I have nothing more pressing to do, I can give it a go in another PR.

**Comment by cameel** [comment]
I mean, it's probably something that @pgebal would pick sooner or later. We just need an issue for it.

File: test/libyul/yulSyntaxTests/blobbasefee_reserved_identifier_post_cancun.yul [hunk 12]
```diff
@@ -0,0 +1,7 @@
+{
+    function blobbasefee() {}
+}
```
Comment (r0qs) [inline test/libyul/yulSyntaxTests/blobbasefee_reserved_identifier_post_cancun.yul:3]:
```suggestion
{
    {
        let blobbasefee := 1
    }

    {
        function blobbasefee() {}
        blobbasefee()
    }
}
```

Maybe we could keep the same format for the tests across all cancun support prs, i.e. declaring variables and functions with the reserved identifier. For instance, see: https://github.com/ethereum/solidity/pull/14779/files#diff-91342678381e6f1b399e20e7bbcee0c7ee82d170010eb5770950a2977d19b9a7

**Review by r0qs — DISMISSED** [review]
LGTM as well, although I have one last comment.

File: test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol
Comment (nikola-matic) [reply test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol:1]: > - `blobbasefee` to the `builtins` test in `FunctionCallGraph.cpp`.

This is not gonna be possible until `cancun` is made the default version, since it just errors out at the analysis stage, and this error is covered by other tests.

> - `environment_access` test in `ViewPureChecker.cpp`.

Same thing as above, however, here I did add the coverage (however pointless at the moment), since it will fail the test once we do move to `cancun`, and then we can get rid of the error, and move it up into the allowed builtins. I wanted to do the same for the `FunctionCallGraph` test, but that would have been too much of a hassle, with having to import `AnalysisFramework.h`, and then refactor.

>- `syntaxTests/viewPureChecker/inline_assembly_instructions_allowed...` group of tests

Done

In addition, @bshastry created https://github.com/ethereum/solidity/pull/14781 for the fuzzer, and I have also created an issue for the [SMT checker](https://github.com/ethereum/solidity/issues/14780) last night.

File: test/libyul/yulSyntaxTests/blobbasefee_reserved_identifier_post_cancun.yul (hunk 12 above)
Comment (nikola-matic) [reply test/libyul/yulSyntaxTests/blobbasefee_reserved_identifier_post_cancun.yul:3]: I tried it - it doesn't matter since it will throw the same parser error as it does now, and stop parsing; also, the error is the same regardless of whether it's a function or a variable declaration, which is kinda weird.

File: test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol
Comment (cameel) [reply test/libsolidity/semanticTests/inlineAssembly/blobbasefee_shanghai_function.sol:1]: Honestly, I expected that this would just work by adding it to those lists. I see it ended up being more complicated. In any case, it's fine the way you did it.

**Review by cameel — APPROVED** [review]
Looks good to merge after squashing.

File: test/libsolidity/ViewPureChecker.cpp [hunk 13]
```diff
@@ -92,6 +92,26 @@ BOOST_AUTO_TEST_CASE(environment_access)
 		TypeError,
 		"\"block.blockhash()\" has been deprecated in favor of \"blockhash()\""
 	);
+
+	std::string baseFeeContract = "contract C { function f() view public { block.basefee; } }";
```
Comment (r0qs) [inline test/libsolidity/ViewPureChecker.cpp:96]: Oh! Shouldn't we do that in another PR for `basefee`? I just noticed that it is also missing `prevrandao` there, so it would be nice to create another PR adding all missing global variables and functions.

Comment (nikola-matic) [reply test/libsolidity/ViewPureChecker.cpp:96]: `prevrandao` is technically covered, since `block.difficulty` is in there - which has the same opcode as `prevrandao`, which means it falls into the same case (in the instruction switch), and then makes a check based on the EVM version as to which one to use.

File: scripts/error_codes.py [hunk 14]
```diff
@@ -202,7 +202,7 @@ def examine_id_coverage(top_dir, source_id_to_file_names, new_ids_only=False):
         "4591", # "There are more than 256 warnings. Ignoring the rest."
                 # Due to 3805, the warning lists look different for different compiler builds.
         "1834", # Unimplemented feature error, as we do not test it anymore via cmdLineTests
-        "5430", # basefee being used in inline assembly for EVMVersion < london
```
Comment (r0qs) [inline scripts/error_codes.py:205]: But why this was replaced? Shouldn't we only add `blobbasefee` there?

Comment (nikola-matic) [reply scripts/error_codes.py:205]: Because of [this change](https://github.com/ethereum/solidity/pull/14755/files#diff-77a7a77f2a2831156ab9f1652dabc5af0858e8ff6739dd3890540ff1d3c5f875). Initially (in the original PR), it was like this, but the inline assembly part was then removed that would yield this error (5430). We've now decided to bring it back, but hardcode it to `berlin`, since the error changes depending on the EVM fork you're running the tests againsts.

File: test/libsolidity/ViewPureChecker.cpp (hunk 13 above)
Comment (cameel) [reply test/libsolidity/ViewPureChecker.cpp:96]: Also, the changes related to `basefee` were small enough that IMO it's fine to put them here. Would not hurt to have them in a separate commit, but that's just a small detail.

File: scripts/error_codes.py (hunk 14 above)
Comment (r0qs) [reply scripts/error_codes.py:205]: Ah ok, thanks for clarifying it;)

File: docs/yul.rst [hunk 15]
```diff
@@ -752,8 +752,8 @@ This document does not want to be a full description of the Ethereum virtual mac
 Please refer to a different document if you are interested in the precise semantics.
 
 Opcodes marked with ``-`` do not return a result and all others return exactly one value.
-Opcodes marked with ``F``, ``H``, ``B``, ``C``, ``I``, ``L`` and ``P`` are present since Frontier,
-Homestead, Byzantium, Constantinople, Istanbul, London or Paris respectively.
+Opcodes marked with ``F``, ``H``, ``B``, ``C``, ``I``, ``L``, ``P`` and ``N`` are present since Frontier,
+Homestead, Byzantium, Constantinople, Istanbul, London, Paris or Cancun respectively.
```
Comment (Saw-mon-and-Natalie) [inline docs/yul.rst:756]: why `N` for Cancun?
